core/src/main/java/org/apache/accumulo/core/classloader/ClassLoaderUtil.java

```diff
@@ -82,6 +82,14 @@ public static ClassLoader getClassLoader(String context) {
     }
   }
 
+  public static boolean isValidContext(String context) {
+    if (context != null && !context.isEmpty()) {
```

Comment:
This can be decided by the factory as well. There's no need to constrain what the factory considers valid.

Reply:
I don't believe that an empty or null context name is valid. I don't think that a user can set that in the shell. I don't think that it can be set via the API either, but I would have to look to be sure. A null or empty context here could indicate a bug.

Reply:
Setting a table property via the shell or API ends up [here](https://github.com/apache/accumulo/blob/2.1/server/manager/src/main/java/org/apache/accumulo/manager/ManagerClientServiceHandler.java#L558). If `context` is null or empty, then there could be an issue with the Accumulo code.

Reply:
If the context property is not set for a particular table, the factory should still get consulted. The empty string is the default value for `table.class.loader.context`, so it most definitely should be valid. The factory should get passed an empty string if the user hasn't overridden it with anything else. It should be up to the factory to handle that according to its own implementation... whether it passes through to the system classloader, or contains a built-in default classloader to return is up to the implementation.

As for the part of the code you point to, I'm not exactly sure what it's doing there... I would think that it should be possible to override a value of `"default"` with a value of `""`, but that code you pointed to seemed to interpret `""` as `unset()`. I'm not sure it's doing the right thing... but that would be a separate issue. Right now, I'm just concerned about what values the factory should allow... and I don't think we need to constrain it to only see non-empty String values as valid, when an empty String is the default, and should be perfectly valid. I'm also of the opinion that `null` should be valid as well. For example, if we had a context classloader factory implementation called `AlwaysReturnSystemClassLoader`, then all context values are valid, and would result in returning the system class loader, with the actual value ignored. That trivial implementation seems like it should be allowed, but it won't work if constraints about validity are imposed outside the SPI implementation.

Reply:
> The empty string is the default value for `table.class.loader.context`, so it most definitely should be valid.

I'm not sure about that. There are no properties of PropertyType.STRING in Property.java whose default value is `null`. They all either have a value, or an empty string. I don't think that empty string is a valid String property value in all cases. I'm sure that in some cases it means the property is not set.

> If the context property is not set for a particular table, the factory should still get consulted.

It's never worked like this. Even in `main` where the VFS stuff has been ripped out ClassLoaderUtil only consults the ContextClassLoaderFactory when the context name is set to a non-empty String. However, I think I have a solution - I pushed the null and empty checks into the factory in a4750ab.

Reply:
> > The empty string is the default value for `table.class.loader.context`, so it most definitely should be valid.
> 
> I'm not sure about that. There are no properties of PropertyType.STRING in Property.java whose default value is `null`. They all either have a value, or an empty string. I don't think that empty string is a valid String property value in all cases. I'm sure that in some cases it means the property is not set.

I don't think we've done a good job in the internal config code of having a way to distinguish between set or not set, or allowing a value at one scope to be overridden by unsetting it at a different scope. This is a shortcoming in the current implementation, but far out of scope to fix here. The fact that `null` isn't used is really more of a byproduct of how commons-configuration works (it's really hard to set it as a value, because it either looks like an empty string, or looks like no entry). In theory, it could be a value, but in practice, you can't really set it. But empty string certainly can be set, and is set as the default in many properties. I don't think we have anything enforcing that `null` can't be the default value, though. If we're going to rely on that, we should probably enforce it.

> 
> > If the context property is not set for a particular table, the factory should still get consulted.
> 
> It's never worked like this. Even in `main` where the VFS stuff has been ripped out ClassLoaderUtil only consults the ContextClassLoaderFactory when the context name is set to a non-empty String.

That is surprising to me. I was thinking it should always be consulted where there is a context possible. But now I'm not so sure what the right behavior should be, because there are too many edge cases and failure scenarios. Trying to put myself in the shoes of a user who has implemented a custom factory, it's getting harder to imagine how they might expect things to behave with respect to their implementation under all these scenarios. I'm leaning towards always consulting the factory, and leaving the "return null" mean "defer to system classloader", but I can also see an argument for "use context classloader factory only if a non-empty context is provided" (which is the current behavior). I'm fine with leaving the current behavior while I think about it some more.

> However, I think I have a solution - I pushed the null and empty checks into the factory in [a4750ab](https://github.com/apache/accumulo/commit/a4750ab12a9694c8b2c0a5590451bdaefa311ed5).

I think pushing these checks into the factory makes sense, and allowing the factory to return null, which causes Accumulo to use its system class loader. However, it's still a little unclear whether contexts that result in returning `null`, and deferring to the system class loader, are "valid" or not. Does "valid" mean "I won't cause an error" or does valid mean "I won't return null"?